File: solutions/python/matching-brackets/1/matching_brackets.py
def is_paired(input_string):
    stack = []
    mapping = {')': '(', ']': '[', '}': '{'}
    
    for char in input_string:
        if char in mapping.values():
            stack.append(char)
        elif char in mapping.keys():
            if not stack or stack.pop() != mapping[char]:
                return False
    
    return not stack

File: solutions/ruby/matching-brackets/1/matching_brackets.rb
class Brackets
  PAIRS = { '[' => ']', '{' => '}', '(' => ')' }

  def self.paired?(str)
    stack = []
    str.each_char do |c|
      if PAIRS.key?(c)
        stack.push(c)
      elsif PAIRS.values.include?(c)
        return false if stack.empty? || PAIRS[stack.pop] != c
      end
    end
    stack.empty?
  end
end
